### PR TITLE
[DependencyInjection] Reword the explanation about index_by and default_index_method

### LINKED ISSUE #18725: Improve `tagged_iterator` documentation
[Documentation about `tagged_iterator`](https://symfony.com/doc/current/service_container/tags.html#tagged-services-with-index) is not easy to understand, because source code have a lot of cases.

I hope it will help to close https://github.com/symfony/symfony/issues/45809 and https://github.com/symfony/symfony/issues/49859

I will try to explain cases. I tested behavior on 5.4 branch.

## `index_by` attribute only

When `index_by` attribute is defined on `tagged_iterator`;
	- Symfony will use this attribute value on tagged service
	- If not defined, it will check the dynamic method `getDefault<CamelCase index_by value>Name` on service
	- If has not method, serviceId will be used 

 Example:

```yaml
# services.yaml
services:
    App\Handler\One:
        class: App\Handler\One
        tags:
            - { name: 'app.handler', handler: 'handler_one' }
    
    App\Handler\Two:
        tags:
            - { name: 'app.handler'}
    
    App\Handler\Three:
        tags:
            - { name: 'app.handler' }
    
    App\HandlerCollection:
        arguments: [!tagged_iterator { tag: 'app.handler', index_by: 'handler' }]
```

```php
final class One
{
}
```

```php
final class Two
{
    public static function getDefaultHandlerName(): string
    {
        return 'handler_two';
    }
}
```
```php
final class Three
{
}
```

When I dump handlers, result is

```
^ array:3 [▼
  "handler_one" => App\Handler\One {#2}
  "App\Handler\Three" => App\Handler\Three {#45}
  "handler_two" => App\Handler\Two {#53}
]
```

## `default_index_method` attribute only

When `default_index_method` attribute is defined on `tagged_iterator`;
	- Symfony will check defined method on service
	- If has not method, serviceId will be used 

 Example:
```yaml
# services.yaml
services:
    App\Handler\One:
        class: App\Handler\One
        tags:
            - { name: 'app.handler'}

    App\Handler\Two:
        tags:
            - { name: 'app.handler'}

    App\HandlerCollection:
        arguments: [!tagged_iterator { tag: 'app.handler', default_index_method: 'getDefaultHandlerName' }]
```

```php
final class One
{
}
```

```php
final class Two
{
    public static function getDefaultHandlerName(): string
    {
        return 'handler_two';
    }
}
```
When I dump handlers, result is

```
^ array:2 [▼
  "App\Handler\One" => App\Handler\One {#3987}
  "handler_two" => App\Handler\Two {#4333}
]
```

## `index_by` and `default_index_method` attribute at same time

When `default_index_method` and `index_by` attributes are defined on `tagged_iterator`;
    - Symfony will use this attribute value on tagged service
    - If not defined, it will check the defined method in `default_index_method`, **method getDefault<CamelCase index_by value>Name is not checked**
    - If has not method, serviceId will be used 

Example:

```yaml
# services.yaml
services:
    App\Handler\One:
        class: App\Handler\One
        tags:
            - { name: 'app.handler', handler: 'handler_one' }

    App\Handler\Two:
        tags:
            - { name: 'app.handler'}

    App\Handler\Three:
        tags:
            - { name: 'app.handler' }

    App\HandlerCollection:
        arguments: [!tagged_iterator { tag: 'app.handler', index_by: 'handler', default_index_method: 'getDefaultFooName' }]

```

```php
final class One
{
    public static function getDefaultFooName(): string
    {
        return 'handler_one_not_applied';
    }
}
```

```php
final class Two
{
    public static function getDefaultFooName(): string
    {
        return 'handler_two';
    }
}
```
```php
final class Three
{
    public static function getDefaultHandlerName(): string
    {
        return 'handler_three';
    }
}
```

When I dump handlers, result is

```
^ array:3 [▼
  "handler_one" => App\Handler\One {#4324}
  "App\Handler\Three" => App\Handler\Three {#4100}
  "handler_two" => App\Handler\Two {#4077}
]
```

## My two cents

I tried to improve documentation on my side, but I didn't found a solution.
I think we should talk about fallbacks, and maybe create hybrid examples like I did here.
Actual doc is confusing because we use same `tagged_iterator` config, but we show them as 2 different ways: 

If someone want to improve or rework this part, I will be happy to help and review !

### PR DESCRIPTION
**Pull request by javiereguiluz**
Fixes #18725.

### CODE REVIEW

**Review by alexandre-daubois — APPROVED** [review]
Great improvements! :smiley:

File: service_container/service_subscribers_locators.rst [hunk 1]
```diff
@@ -507,12 +507,15 @@ will share identical locators among all the services referencing them::
 Indexing the Collection of Services
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-Services passed to the service locator can define their own index using an
-arbitrary attribute whose name is defined as ``index_by`` in the service locator.
+By default, services passed to the service locator are indexed using their service
+IDs. You can change this bahavior with two options of the tagged locator (``index_by``
```
Comment (alamirault) [inline service_container/service_subscribers_locators.rst]:
```suggestion
IDs. You can change this behavior with two options of the tagged locator (``index_by``
```

File: service_container/service_subscribers_locators.rst [hunk 2]
```diff
@@ -608,31 +612,25 @@ Inside this locator you can retrieve services by index using the value of the
     {
         public function __construct(ServiceLocator $locator)
         {
+            // this value is defined in the `key` option of the servive
```
Comment (alamirault) [inline service_container/service_subscribers_locators.rst]:
```suggestion
            // this value is defined in the `key` option of the service
```

**Review by alamirault — CHANGES_REQUESTED** [review]
I really love your rewrite !

**Comment by javiereguiluz** [comment]
Thanks for the review! I've made the same changes to the other article. Cheers!

**Review by alamirault — APPROVED** [review]

**Pull request merged**